File: spec/javascript/server_spec.js
var jstest = require("jstest").Test

var Engine = require("../../src/engines/proxy"),
    Server = require("../../src/protocol/server")

jstest.describe("Server", function() { with(this) {
  before(function() { with(this) {
    this.engine = {}
    stub(Engine, "get").returns(engine)
    this.server = new Server()
  }})

  describe("#process", function() { with(this) {
    before(function() { with(this) {
      this.handshake   = {channel: "/meta/handshake",   data: "handshake"}
      this.connect     = {channel: "/meta/connect",     data: "connect"}
      this.disconnect  = {channel: "/meta/disconnect",  data: "disconnect"}
      this.subscribe   = {channel: "/meta/subscribe",   data: "subscribe"}
      this.unsubscribe = {channel: "/meta/unsubscribe", data: "unsubscribe"}
      this.publish     = {channel: "/some/channel",     data: "publish"}

      stub(engine, "interval", 0)
      stub(engine, "timeout", 60)
    }})

    it("returns an empty response for no messages", function() { with(this) {
      var response = null
      server.process([], false, function(r) { response = r})
      assertEqual( [], response )
    }})

    it("ignores invalid messages", function() { with(this) {
      var response = null
      server.process([{}, {channel: "invalid"}], false, function(r) { response = r})
      assertEqual([
        { successful: false,
          error:      "402:data:Missing required parameter"
        },
        { channel:    "invalid",
          successful: false,
          error:      "402:data:Missing required parameter"
        }
      ], response)
    }})

    it("rejects unknown meta channels", function() { with(this) {
      var response = null
      server.process([{channel: "/meta/p"}], false, function(r) { response = r })
      assertEqual([
        { channel:    "/meta/p",
          successful: false,
          error:      "403:/meta/p:Forbidden channel"
        }
      ], response)
    }})

    it("routes single messages to appropriate handlers", function() { with(this) {
      expect(server, "handshake").given(handshake, false).yielding([{}])
      server.process(handshake, false, function() {})
    }})

    it("routes a list of messages to appropriate handlers", function() { with(this) {
      expect(server, "handshake").given(handshake, false).yielding([{}])
      expect(server, "connect").given(connect, false).yielding([{}])
      expect(server, "disconnect").given(disconnect, false).yielding([{}])
      expect(server, "subscribe").given(subscribe, false).yielding([{}])
      expect(server, "unsubscribe").given(unsubscribe, false).yielding([{}])

      expect(engine, "publish").given(handshake).exactly(0)
      expect(engine, "publish").given(connect).exactly(0)
      expect(engine, "publish").given(disconnect).exactly(0)
      expect(engine, "publish").given(subscribe).exactly(0)
      expect(engine, "publish").given(unsubscribe).exactly(0)

      expect(engine, "publish").given(publish)

      server.process([handshake, connect, disconnect, subscribe, unsubscribe, publish], false, function() {})
    }})

    describe("handshaking", function() { with(this) {
      before(function() { with(this) {
        expect(server, "handshake").given(handshake, false).yielding([{channel: "/meta/handshake", successful: true}])
      }})

      it("returns the handshake response with advice", function() { with(this) {
        server.process(handshake, false, function(response) {
          assertEqual([
              { channel: "/meta/handshake",
                successful: true,
                advice: {reconnect: "retry", interval: 0, timeout: 60000}
              }
            ], response)
        })
      }})
    }})

    describe("connecting for messages", function() { with(this) {
      before(function() { with(this) {
        this.messages = [{channel: "/a"}, {channel: "/b"}]
        expect(server, "connect").given(connect, false).yielding([messages])
      }})

      it("returns the new messages", function() { with(this) {
        server.process(connect, false, function(response) {
          assertEqual( messages, response )
        })
      }})
    }})
  }})
}})

File: spec/ruby/server_spec.rb
require "spec_helper"

describe Faye::Server do
  let(:engine) { double "engine" }
  let(:server) { Faye::Server.new }

  before do
    Faye::Engine.stub(:get).and_return engine
  end

  describe :process do
    let(:handshake)   {{"channel" => "/meta/handshake",   "data" => "handshake"  }}
    let(:connect)     {{"channel" => "/meta/connect",     "data" => "connect"    }}
    let(:disconnect)  {{"channel" => "/meta/disconnect",  "data" => "disconnect" }}
    let(:subscribe)   {{"channel" => "/meta/subscribe",   "data" => "subscribe"  }}
    let(:unsubscribe) {{"channel" => "/meta/unsubscribe", "data" => "unsubscribe"}}
    let(:publish)     {{"channel" => "/some/channel",     "data" => "publish"    }}

    before do
      engine.stub(:interval).and_return(0)
      engine.stub(:timeout).and_return(60)
    end

    it "returns an empty response for no messages" do
      response = nil
      server.process([], false) { |r| response = r }
      response.should == []
    end

    it "ignores invalid messages" do
      response = nil
      server.process([{}, {"channel" => "invalid"}], false) { |r| response = r }
      response.should == [
        { "successful"  => false,
          "error"       => "405::Invalid channel"
        },
        { "channel"     => "invalid",
          "successful"  => false,
          "error"       => "405:invalid:Invalid channel"
        }
      ]
    end

    it "rejects unknown meta channels" do
      response = nil
      server.process([{"channel" => "/meta/p"}], false) { |r| response = r }
      response.should == [
        { "channel"     => "/meta/p",
          "successful"  => false,
          "error"       => "403:/meta/p:Forbidden channel"
        }
      ]
    end

    it "routes single messages to appropriate handlers" do
      server.should_receive(:handshake).with(handshake, false)
      server.process(handshake, false) {}
    end

    it "routes a list of messages to appropriate handlers" do
      server.should_receive(:handshake).with(handshake, false)
      server.should_receive(:connect).with(connect, false)
      server.should_receive(:disconnect).with(disconnect, false)
      server.should_receive(:subscribe).with(subscribe, false)
      server.should_receive(:unsubscribe).with(unsubscribe, false)

      engine.should_not_receive(:publish).with(handshake)
      engine.should_not_receive(:publish).with(connect)
      engine.should_not_receive(:publish).with(disconnect)
      engine.should_not_receive(:publish).with(subscribe)
      engine.should_not_receive(:publish).with(unsubscribe)
      engine.should_receive(:publish).with(publish)

      server.process([handshake, connect, disconnect, subscribe, unsubscribe, publish], false)
    end

    describe "handshaking" do
      before do
        response = {"channel" => "/meta/handshake", "successful" => true}
        server.should_receive(:handshake).with(handshake, false).and_yield(response)
      end

      it "returns the handshake response with advice" do
        server.process(handshake, false) do |response|
          response.should == [
            { "channel" => "/meta/handshake",
              "successful" => true,
              "advice" => {"reconnect" => "retry", "interval" => 0, "timeout" => 60000}
            }
          ]
        end
      end
    end

    describe "connecting for messages" do
      let(:messages) { [{"channel" => "/a"}, {"channel" => "/b"}] }

      before do
        server.should_receive(:connect).with(connect, false).and_yield(messages)
      end

      it "returns the new messages" do
        server.process(connect, false) { |r| r.should == messages }
      end
    end
  end
end
